### ISSUE TITLE
Loom: VM Pinning support

### ISSUE THREAD

**Issue by tajila** [eclipse-openj9/openj9#15174]
Depends on https://github.com/eclipse-openj9/openj9/issues/15172

When a Virtual thread enters a synch block (or method) it is pinned until it exits. Likewise for native functions and Object.wait().

Part 1: Basic Pinning support
- JNI callout increment counter, decrement when callout frame is popped
- monenter increment counter, vice versa on monexit (same for sync methods)
- Object.wait inc counter on entry, vice versa on exit

Part 2: Exception Throws
On exception throw, 
- track all frames that need to dec/inc the pin state counter as well
- real exception throws in JIT code handled by VM runtime helpers

**Comment by babsingh** [comment]
Part 2: the goal is to decrement the pin count while exiting a sync block for an exception.

**Comment by gacholio** [comment]
Look in ObjectMonitor.hpp and monhelpers.c - these are the things that eventually lock/unlock an object, so they may as well manage the lock pin count.

**Comment by tajila** [comment]
@babsingh Note: you'll also need to implement 

```
public static native void pin();

public static native void unpin();

public static boolean isPinned(ContinuationScope scope);

```

For pin/unpin youll need to test the RI behaviour to determine if it inc/dec the counter or if it completely ignores it.

**Comment by babsingh** [comment]
> Object.wait inc counter on entry, vice versa on exit

`Object.wait` should only be invoked inside the scope of `monenter` and `monexit`. Do we need to track `Object.wait` separately or just incrementing the counter once in `monenter` should be sufficient to account for it?

```
monenter
...
Object.wait
...
monexit
```

**Comment by babsingh** [comment]
> JNI callout increment counter, decrement when callout frame is popped

Do we need to track all the below frame types with a single counter?
```
// These two should cover the JNI callouts
#define J9SF_FRAME_TYPE_JNI_NATIVE_METHOD  7
#define J9SF_FRAME_TYPE_JIT_JNI_CALLOUT  6

// Does this also need to be tracked?
#define J9SF_FRAME_TYPE_NATIVE_METHOD  3
```

**Comment by tajila** [comment]
Youll need to do `J9SF_FRAME_TYPE_JNI_NATIVE_METHOD` for sure. Also the new style FFI Lnker callouts that @ChengJin01 is working on.

@gacholio is `J9SF_FRAME_TYPE_JIT_JNI_CALLOUT` used for directJNI calls by the JIT?

I'm not sure `J9SF_FRAME_TYPE_NATIVE_METHOD` will need to be addressed in all cases, but any case where that can potentialy call back into the JVM (class init, exception constructors, resolve methods, etc.) then it should also pin

**Comment by gacholio** [comment]
I don't understand the question - are you proposing not keeping a counter for callout and walking the stack to determine it?

**Comment by tajila** [comment]
> I don't understand the question - are you proposing not keeping a counter for callout and walking the stack to determine it?

No the question is "where does this counter need to be inc/dec?"

**Comment by tajila** [comment]
In other words, do we need to increment the counter every time we push a `J9SF_FRAME_TYPE_JNI_NATIVE_METHOD` frame? I believe the answer is yes.

Do we need to increment the counter every time we push a `J9SF_FRAME_TYPE_NATIVE_METHOD` frame? I think the answer here is sometimes.

**Comment by gacholio** [comment]
All callouts - interpreter FFI, direct JNI from JIT, new direct C callout, library load callbacks (JNI_OnLoad, Agent_OnLoad, Agent_OnAttach and matching unload calls).

**Comment by gacholio** [comment]
Yes, the counters must be managed for every call (which is why it's a counter and not a boolean).

**Comment by gacholio** [comment]
We don't need the counter for INLs that don't release VM access, I think, but the simplicity of managing the counter in the frame build is appealing.

**Comment by tajila** [comment]
@babsingh If one calls Continuation::yield and the Vthread is pinned what is the result?

**Comment by babsingh** [comment]
> If one calls Continuation::yield and the Vthread is pinned what is the result?

On the Java side, `Continuation::yield` should throw an `IllegalStateException`; message: pinned reason `{NATIVE, MONITOR, CRITICAL_SECTION}`.

On the native side, the return code should match the values of `Pinned` enum.

**Comment by tajila** [comment]
> On the native side, the return code should match the values of Pinned enum.

Which return code are you talking about?

**Comment by babsingh** [comment]
> Which return code are you talking about?

See https://github.com/eclipse-openj9/openj9/pull/15586, which contains the described changes.

**Comment by tajila** [comment]
Thanks

**Comment by fengxue-IS** [comment]
Closing this issue as workarounds in #15586 has been addressed by #15678
